src/lib/operators/aggregate_hash.hpp

```diff
@@ -81,7 +83,7 @@ using AggregateKeyEntry = uint64_t;
 struct EmptyAggregateKey {};
 
 template <typename AggregateKey>
-using AggregateKeys = std::vector<AggregateKey>;
+using AggregateKeys = std::conditional_t<std::is_same_v<AggregateKey, boost::container::small_vector<AggregateKeyEntry, 4>>, std::vector<AggregateKey>, uninitialized_vector<AggregateKey>>;
```

Comment:
(6) Use of uninitialized_vector

Reply:
https://github.com/hyrise/hyrise/compare/61bd7a4..19dac4e

Uninitialized_vector in combination with small_vector is slightly beneficial

```
+----------++----------+---------+--------++----------+----------+--------+---------+
| Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
|          ||      old |     new |        ||      old |      new |        |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
| TPC-H 01 ||   1651.5 |  1701.2 |   +3%  ||     0.61 |     0.59 |   -3%  |  0.0866 |
| TPC-H 02 ||     11.8 |    11.6 |   -2%  ||    84.50 |    86.33 |   +2%  |  0.0783 |
| TPC-H 03 ||    178.4 |   164.3 |   -8%  ||     5.61 |     6.09 |   +9%  |  0.0000 |
| TPC-H 04 ||    152.0 |   151.6 |   -0%  ||     6.58 |     6.60 |   +0%  |  0.0146 |
| TPC-H 05 ||    470.9 |   465.3 |   -1%  ||     2.12 |     2.15 |   +1%  |  0.0064 |
| TPC-H 06 ||      7.9 |     7.8 |   -1%  ||   126.54 |   127.55 |   +1%  |  0.0022 |
| TPC-H 07 ||    126.2 |   124.5 |   -1%  ||     7.93 |     8.03 |   +1%  |  0.2119 |
| TPC-H 08 ||    156.3 |   156.9 |   +0%  ||     6.40 |     6.37 |   -0%  |  0.4357 |
| TPC-H 09 ||   1013.1 |  1000.2 |   -1%  ||     0.99 |     1.00 |   +1%  |  0.0017 |
| TPC-H 10 ||    511.9 |   509.3 |   -1%  ||     1.95 |     1.96 |   +1%  |  0.6105 |
| TPC-H 11 ||     16.5 |    16.4 |   -0%  ||    60.71 |    60.78 |   +0%  |  0.4820 |
| TPC-H 12 ||     90.0 |    88.1 |   -2%  ||    11.11 |    11.35 |   +2%  |  0.0000 |
| TPC-H 13 ||    705.9 |   709.5 |   +1%  ||     1.42 |     1.41 |   -1%  |  0.0003 |
| TPC-H 14 ||     38.8 |    37.7 |   -3%  ||    25.77 |    26.52 |   +3%  |  0.0000 |
| TPC-H 15 ||     17.6 |    17.5 |   -0%  ||    56.89 |    57.01 |   +0%  |  0.0440 |
| TPC-H 16 ||    178.3 |   117.3 |  -34%  ||     5.61 |     8.53 |  +52%  |  0.0000 |
| TPC-H 17 ||     54.4 |    54.0 |   -1%  ||    18.39 |    18.53 |   +1%  |  0.0008 |
| TPC-H 18 ||   1312.0 |  1289.2 |   -2%  ||     0.76 |     0.78 |   +2%  |  0.0005 |
| TPC-H 19 ||     75.3 |    74.9 |   -0%  ||    13.28 |    13.35 |   +0%  |  0.0001 |
| TPC-H 20 ||     31.7 |    31.6 |   -0%  ||    31.54 |    31.65 |   +0%  |  0.2689 |
| TPC-H 21 ||    822.2 |   823.4 |   +0%  ||     1.22 |     1.21 |   -0%  |  0.8250 |
| TPC-H 22 ||    126.3 |   126.3 |   +0%  ||     7.92 |     7.92 |   -0%  |  0.8783 |
+----------++----------+---------+--------++----------+----------+--------+---------+
| Sum      ||   7748.8 |  7678.7 |   -1%  ||          |          |        |         |
| Geomean  ||          |         |        ||          |          |   +3%  |         |
+----------++----------+---------+--------++----------+----------+--------+---------+
```

Reply:
Hmm, it's weird that Q16 shows up again. I wonder if this is more of an alignment issue than a code change issue.

Reply:
Ah, misread it.

Reply:
Nemea:

```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | TPC-H 01 ||    821.3 |   812.6 |   -1%  ||     1.22 |     1.23 |   +1%  |  0.0985 |
 | TPC-H 02 ||      4.6 |     4.5 |   -1%  ||   218.39 |   219.76 |   +1%  |  0.4180 |
 | TPC-H 03 ||     81.4 |    80.9 |   -1%  ||    12.28 |    12.36 |   +1%  |  0.0000 |
 | TPC-H 04 ||     65.9 |    65.4 |   -1%  ||    15.18 |    15.29 |   +1%  |  0.0000 |
 | TPC-H 05 ||    223.3 |   222.7 |   -0%  ||     4.48 |     4.49 |   +0%  |  0.5491 |
 | TPC-H 06 ||      3.2 |     3.2 |   -1%  ||   310.77 |   312.94 |   +1%  |  0.0000 |
 | TPC-H 07 ||     57.0 |    56.8 |   -0%  ||    17.54 |    17.61 |   +0%  |  0.6921 |
 | TPC-H 08 ||     71.7 |    72.0 |   +0%  ||    13.95 |    13.88 |   -0%  |  0.3212 |
 | TPC-H 09 ||    508.6 |   505.8 |   -1%  ||     1.97 |     1.98 |   +1%  |  0.1012 |
 | TPC-H 10 ||    249.0 |   250.5 |   +1%  ||     4.02 |     3.99 |   -1%  |  0.5634 |
 | TPC-H 11 ||      8.3 |     8.2 |   -1%  ||   120.62 |   122.03 |   +1%  |  0.0000 |
 | TPC-H 12 ||     42.4 |    42.3 |   -0%  ||    23.60 |    23.66 |   +0%  |  0.0333 |
 | TPC-H 13 ||    354.9 |   352.5 |   -1%  ||     2.82 |     2.84 |   +1%  |  0.0000 |
 | TPC-H 14 ||     21.2 |    21.1 |   -1%  ||    47.06 |    47.40 |   +1%  |  0.0000 |
 | TPC-H 15 ||      7.7 |     7.7 |   -1%  ||   129.04 |   130.50 |   +1%  |  0.0000 |
 | TPC-H 16 ||     91.2 |    90.6 |   -1%  ||    10.96 |    11.04 |   +1%  |  0.0051 |
 | TPC-H 17 ||     16.4 |    16.3 |   -1%  ||    60.87 |    61.52 |   +1%  |  0.0000 |
+| TPC-H 18 ||    731.0 |   679.0 |   -7%  ||     1.37 |     1.47 |   +8%  |  0.0000 |
 | TPC-H 19 ||     27.8 |    27.8 |   -0%  ||    35.95 |    36.02 |   +0%  |  0.0322 |
 | TPC-H 20 ||     15.7 |    15.6 |   -1%  ||    63.80 |    64.19 |   +1%  |  0.0185 |
 | TPC-H 21 ||    414.8 |   410.7 |   -1%  ||     2.41 |     2.43 |   +1%  |  0.0430 |
 | TPC-H 22 ||     50.0 |    49.3 |   -1%  ||    20.01 |    20.26 |   +1%  |  0.0000 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Sum      ||   3867.4 |  3795.4 |   -2%  ||          |          |        |         |
 | Geomean  ||          |         |        ||          |          |   +1%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```

This is disturbing. Knowing that we could get +50% in one query but not knowing how to get there is worse than blissful ignorance. I feel like this could be a really interesting research project though.

Reply:
```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | TPC-H 01 ||    810.9 |   819.2 |   +1%  ||     1.23 |     1.22 |   -1%  |  0.0989 |
+| TPC-H 02 ||      5.2 |     4.8 |   -7%  ||   193.44 |   207.77 |   +7%  |  0.0000 |
+| TPC-H 03 ||     83.8 |    75.7 |  -10%  ||    11.93 |    13.21 |  +11%  |  0.0000 |
 | TPC-H 04 ||     68.4 |    68.5 |   +0%  ||    14.62 |    14.59 |   -0%  |  0.0052 |
 | TPC-H 05 ||    232.3 |   230.3 |   -1%  ||     4.30 |     4.34 |   +1%  |  0.0262 |
+| TPC-H 06 ||      3.5 |     3.2 |  -10%  ||   282.34 |   313.47 |  +11%  |  0.0000 |
 | TPC-H 07 ||     58.5 |    58.5 |   -0%  ||    17.08 |    17.08 |   +0%  |  0.9956 |
 | TPC-H 08 ||     76.5 |    76.4 |   -0%  ||    13.07 |    13.10 |   +0%  |  0.6893 |
 | TPC-H 09 ||    517.0 |   515.8 |   -0%  ||     1.93 |     1.94 |   +0%  |  0.5216 |
 | TPC-H 10 ||    262.4 |   253.7 |   -3%  ||     3.81 |     3.94 |   +3%  |  0.0024 |
 | TPC-H 11 ||      8.9 |     8.6 |   -3%  ||   112.24 |   115.82 |   +3%  |  0.0000 |
 | TPC-H 12 ||     43.6 |    43.1 |   -1%  ||    22.96 |    23.19 |   +1%  |  0.0000 |
 | TPC-H 13 ||    351.5 |   355.8 |   +1%  ||     2.85 |     2.81 |   -1%  |  0.0000 |
 | TPC-H 14 ||     21.1 |    21.0 |   -1%  ||    47.39 |    47.69 |   +1%  |  0.0000 |
 | TPC-H 15 ||      7.7 |     7.7 |   -0%  ||   130.28 |   130.30 |   +0%  |  0.8729 |
+| TPC-H 16 ||     93.8 |    60.6 |  -35%  ||    10.66 |    16.50 |  +55%  |  0.0000 |
 | TPC-H 17 ||     19.5 |    19.2 |   -1%  ||    51.38 |    51.98 |   +1%  |  0.0000 |
 | TPC-H 18 ||    697.6 |   715.6 |   +3%  ||     1.43 |     1.40 |   -3%  |  0.0406 |
 | TPC-H 19 ||     31.4 |    30.8 |   -2%  ||    31.89 |    32.48 |   +2%  |  0.0000 |
 | TPC-H 20 ||     16.5 |    16.5 |   -0%  ||    60.63 |    60.71 |   +0%  |  0.6009 |
 | TPC-H 21 ||    420.9 |   420.7 |   -0%  ||     2.38 |     2.38 |   +0%  |  0.9389 |
 | TPC-H 22 ||     50.5 |    50.5 |   +0%  ||    19.79 |    19.78 |   -0%  |  0.1865 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Sum      ||   3881.3 |  3856.4 |   -1%  ||          |          |        |         |
 | Geomean  ||          |         |        ||          |          |   +4%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```

Let's make a PR out of this and the removal of the `std::array` shortcut.